File: backend/app/schemas/notifications.py
import uuid
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class NotificationType(str, Enum):
    batch_done = "batch_done"
    batch_failed = "batch_failed"
    app_update = "app_update"
    share_invite = "share_invite"


class AnnouncementSeverity(str, Enum):
    info = "info"
    warning = "warning"
    critical = "critical"


class NotificationRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    type: NotificationType
    title: str
    body: str
    is_read: bool
    link_url: str | None = None
    data: dict[str, Any] | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class NotificationListResponse(BaseModel):
    total: int = Field(description="Total number of notifications matching the current filters.")
    next_cursor: str | None = Field(
        default=None,
        description="Opaque cursor for fetching the next page. Keep filters and sort parameters unchanged between requests.",
    )
    has_more: bool = Field(description="Whether there are additional items after this page.")
    page_size: int = Field(description="Number of items returned per page.")
    items: list[NotificationRead] = Field(description="Notifications returned for the current page.")


class AppAnnouncementCreate(BaseModel):
    version: str | None = Field(default=None, max_length=64)
    title: str = Field(min_length=1, max_length=255)
    message: str = Field(min_length=1)
    severity: AnnouncementSeverity = AnnouncementSeverity.info
    starts_at: datetime | None = None
    ends_at: datetime | None = None


class AppAnnouncementRead(BaseModel):
    id: uuid.UUID
    version: str | None = None
    title: str
    message: str
    severity: AnnouncementSeverity
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/app/models/notifications.ts
import { CursorPage } from './common';

export enum NotificationType {
  BATCH_DONE = 'batch_done',
  BATCH_FAILED = 'batch_failed',
  APP_UPDATE = 'app_update',
  SHARE_INVITE = 'share_invite',
  METADATA_REVIEW = 'metadata_review',
  WELCOME = 'welcome',
}

export enum AnnouncementSeverity {
  INFO = 'info',
  WARNING = 'warning',
  CRITICAL = 'critical',
}

export interface ShareInviteNotificationData {
  album_id: string;
  album_name: string;
  role: 'viewer' | 'editor';
  invited_by_user_id: string;
  invited_by_username: string;
  invite_status: 'pending' | 'accepted' | 'rejected';
  invite_id: string;
}

export interface AppUpdateNotificationData {
  announcement_id: string;
  severity: AnnouncementSeverity;
  version: string | null;
  starts_at: string | null;
  ends_at: string | null;
}

export interface MetadataReviewNotificationData {
  latest_batch_id: string;
  review_batch_ids: string[];
  unresolved_count: number;
  dismiss_signature: string;
}

export type NotificationData =
  | ShareInviteNotificationData
  | AppUpdateNotificationData
  | MetadataReviewNotificationData
  | Record<string, unknown>;

export interface NotificationRead {
  id: string;
  user_id: string;
  type: NotificationType;
  title: string;
  body: string;
  is_read: boolean;
  link_url: string | null;
  data: NotificationData | null;
  created_at: string;
}

export interface AppAnnouncementRead {
  id: string;
  version: string | null;
  title: string;
  message: string;
  severity: AnnouncementSeverity;
  starts_at: string | null;
  ends_at: string | null;
  is_active: boolean;
  created_at: string;
}

export interface AppAnnouncementCreate {
  version?: string | null;
  title: string;
  message: string;
  severity?: AnnouncementSeverity;
  starts_at?: string | null;
  ends_at?: string | null;
}

export type NotificationListResponse = CursorPage<NotificationRead>;
